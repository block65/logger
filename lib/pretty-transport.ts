import {
  bgRed,
  bgYellow,
  black,
  blue,
  bold,
  gray,
  green,
  red,
  whiteBright,
} from 'colorette';
import { once } from 'events';
import build from 'pino-abstract-transport';
import SonicBoom, { SonicBoomOpts } from 'sonic-boom';
import { isatty } from 'tty';
import util from 'util';
import { LogDescriptor } from './types.js';

export interface PrettyTransportOptions {
  destination?: number | SonicBoomOpts['dest'];
}

function formatLevel(level: number): string {
  switch (level) {
    case 60: //fatal
      return whiteBright(bgRed(bold('FATAL')));
    case 50: //error
      return red('ERROR');
    case 40: // warn
      return bgYellow(black(bold('WARN')));
    case 30: //info
      return blue('INFO');
    case 20: //debug
      return green('DEBUG');
    case 10: //trace
      return gray('TRACE');
    default:
      return gray('SILENT');
  }
}

export function createPrettifier(/*options?: unknown*/) {
  return (log: LogDescriptor): string => {
    const { level, msg = '', time, name, hostname, pid, ...rest } = log;

    if (msg?.startsWith('pino.final with prettyPrint')) {
      return '';
    }

    const formattedName = name ? `(${name})` : '';
    const formattedMsg = msg && ' ' + whiteBright(msg);

    const formattedRest =
      Object.keys(rest).length > 0
        ? ' ' +
          util.inspect(rest, {
            colors:
              'FORCE_COLOR' in process.env ||
              (isatty(1) &&
                process.env.TERM !== 'dumb' &&
                !('NO_COLOR' in process.env)),
            compact: true,
            sorted: true,
          })
        : '';

    return `${gray(new Date(time).toJSON())} ${formatLevel(level).padEnd(
      6,
      ' ',
    )}${formattedName} ${formattedMsg}${formattedRest}\n`;
  };
}

export default async function (options?: PrettyTransportOptions) {
  const prettifier = createPrettifier();
  // SonicBoom is necessary to avoid loops with the main thread.
  // It is the same of pino.destination().
  const destination = new SonicBoom({
    fd:
      typeof options?.destination === 'number'
        ? options?.destination
        : undefined,
    dest:
      typeof options?.destination !== 'number'
        ? options?.destination?.toString()
        : undefined,
    sync: true,
  });

  await once(destination, 'ready');

  return build(
    async function (source) {
      for await (let obj of source) {
        const toDrain = !destination.write(prettifier(obj) + '\n');
        // This block will handle backpressure
        if (toDrain) {
          await once(destination, 'drain');
        }
      }
    },
    {
      async close(err?: Error) {
        if (err) {
          console.warn(err);
        }
        destination.end();

        await once(destination, 'close');
      },
    },
  );
}
