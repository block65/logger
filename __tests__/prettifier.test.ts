import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { createPrettifier } from '../lib/pretty-transport.js';

process.env.NO_COLOR = 'true';

describe('Prettifier', () => {
  const OLD_ENV = process.env;
  beforeEach(() => {
    jest.resetModules();
    process.env = { ...OLD_ENV };
  });

  test('Color', async () => {
    process.env.NO_COLOR = 'true';

    const prettifier = createPrettifier();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
      }),
    ).toMatchSnapshot();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });

  test('With colour', async () => {
    process.env.FORCE_COLOR = 'true';

    const prettifier = createPrettifier();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
      }),
    ).toMatchSnapshot();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });
});
