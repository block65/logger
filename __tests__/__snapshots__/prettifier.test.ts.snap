// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prettifier Color 1`] = `
"[90m1970-01-15T06:56:07.890Z[39m [32mDEBUG[39m  [97mkek[39m
"
`;

exports[`Prettifier Color 2`] = `
"[90m1970-01-15T06:56:07.890Z[39m [32mDEBUG[39m  [97mkek[39m { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
"
`;

exports[`Prettifier With colour 1`] = `
"[90m1970-01-15T06:56:07.890Z[39m [32mDEBUG[39m  [97mkek[39m
"
`;

exports[`Prettifier With colour 2`] = `
"[90m1970-01-15T06:56:07.890Z[39m [32mDEBUG[39m  [97mkek[39m { random: { arr: [ [33m1[39m, [33m2[39m, [33m3[39m ], bool: [33mtrue[39m, number: [33m12[39m, str: [32m'hello'[39m } }
"
`;
